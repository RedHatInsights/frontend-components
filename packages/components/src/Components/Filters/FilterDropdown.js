import React, { Component } from 'react';
import { Dropdown, DropdownToggle } from '@patternfly/react-core';
import PropTypes from 'prop-types';
import FilterInput from './FilterInput.js';

class FilterDropdown extends Component {
    state = {
        isOpen: false
    };

    addRemoveFilters = (selectedValue, filterName, type, isChecked) => {
        switch (type) {
            case 'checkbox':
                isChecked ? this.props.addFilter(filterName, selectedValue, type) : this.props.removeFilter(filterName, selectedValue);
                break;
            case 'radio':
                this.props.addFilter(filterName, selectedValue, type);
                break;
        }
    };

    onToggle = isOpen => {
        this.setState({
            isOpen
        });
    };

    render() {
        const { hideCategories, filters, filterCategories, label } = this.props;
        const { isOpen } = this.state;

        return (
            <Dropdown
                className="ins-c-filter__dropdown"
                onSelect={ this.onSelect }
                toggle={ <DropdownToggle onToggle={ this.onToggle }>{ label }</DropdownToggle> } isOpen={ isOpen }
            >
                <div className='pf-c-dropdown__menu-item'>
                    { filterCategories.map(
                        (data, index) =>
                            !hideCategories.includes(data.urlParam) && (
                                <div key={ `${data.urlParam}${index}` } className="filterTitle">
                                    { data.title }
                                    { data.values.map((item, key) => (
                                        <FilterInput
                                            key={ `check${index}${key}` }
                                            aria-label={ item.label }
                                            id={ `${data.urlParam}${key}` }
                                            label={ item.label }
                                            addRemoveFilters={ this.addRemoveFilters }
                                            param={ data.urlParam }
                                            type={ data.type }
                                            value={ item.value }
                                            filters={ filters }
                                        />
                                    )) }
                                    { index !== filterCategories.length - 1 && <br /> }
                                </div>
                            )
                    ) }
                </div>
            </Dropdown>
        );
    }
}

FilterDropdown.propTypes = {
    addFilter: PropTypes.func,
    removeFilter: PropTypes.func,
    hideCategories: PropTypes.array,
    filters: PropTypes.object,
    filterCategories: PropTypes.arrayOf(
        PropTypes.shape({
            title: PropTypes.string,
            type: PropTypes.type,
            urlParam: PropTypes.string,
            values: PropTypes.arrayOf(
                PropTypes.shape({
                    label: PropTypes.node,
                    value: PropTypes.any
                })
            )
        })
    ),
    label: PropTypes.string
};

FilterDropdown.defaultProps = {
    addFilter: Function.prototype,
    removeFilter: Function.prototype,
    hideCategories: [],
    filters: {},
    label: "Filters"
};

export default FilterDropdown;
