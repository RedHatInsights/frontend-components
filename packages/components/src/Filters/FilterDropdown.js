import React, { Component } from 'react';
import { Dropdown, DropdownToggle, Level } from '@patternfly/react-core';
import { getDefaultOUIAId } from '@patternfly/react-core/';
import PropTypes from 'prop-types';
import FilterInput from './FilterInput.js';
import './filter-dropdown.scss';

class FilterDropdown extends Component {
  state = {
    isOpen: false,
    ouiaStateId: getDefaultOUIAId('RHI/FilterDropdown'),
  };

  addRemoveFilters = (selectedValue, filterName, type, isChecked) => {
    switch (type) {
      case 'checkbox':
        isChecked ? this.props.addFilter(filterName, selectedValue, type) : this.props.removeFilter(filterName, selectedValue);
        break;
      case 'radio':
        this.props.addFilter(filterName, selectedValue, type);
        break;
    }
  };

  onToggle = (isOpen) => {
    this.setState({
      isOpen,
    });
  };

  render() {
    const { hideCategories, filters, filterCategories, label, ouiaId, ouiaSafe } = this.props;
    const { isOpen, ouiaStateId } = this.state;
    const ouiaFinalId = ouiaId !== undefined ? ouiaId : ouiaStateId;

    return (
      <Dropdown
        className="ins-c-filter__dropdown"
        onSelect={this.onSelect}
        toggle={
          <DropdownToggle ouiaId={ouiaFinalId} ouiaSafe={ouiaSafe} onToggle={this.onToggle}>
            {label}
          </DropdownToggle>
        }
        isOpen={isOpen}
        ouiaId={ouiaFinalId}
        ouiaSafe={ouiaSafe}
      >
        <div className="pf-c-dropdown__menu-item">
          {filterCategories.map(
            (data, index) =>
              !hideCategories.includes(data.urlParam) && (
                <Level key={`${data.urlParam}${index}`}>
                  {data.title}
                  {data.values.map((item, key) => (
                    <FilterInput
                      key={`check${index}${key}`}
                      aria-label={item.label}
                      id={`${data.urlParam}${key}`}
                      label={item.label}
                      addRemoveFilters={this.addRemoveFilters}
                      param={data.urlParam}
                      type={data.type}
                      value={item.value}
                      filters={filters}
                    />
                  ))}
                  {index !== filterCategories.length - 1 && <br />}
                </Level>
              )
          )}
        </div>
      </Dropdown>
    );
  }
}

FilterDropdown.propTypes = {
  addFilter: PropTypes.func,
  removeFilter: PropTypes.func,
  hideCategories: PropTypes.array,
  filters: PropTypes.object,
  filterCategories: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.string,
      type: PropTypes.type,
      urlParam: PropTypes.string,
      values: PropTypes.arrayOf(
        PropTypes.shape({
          label: PropTypes.node,
          value: PropTypes.any,
        })
      ),
    })
  ),
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  ouiaId: PropTypes.string,
  ouiaSafe: PropTypes.bool,
};

FilterDropdown.defaultProps = {
  addFilter: Function.prototype,
  removeFilter: Function.prototype,
  hideCategories: [],
  filters: {},
  label: 'Filters',
  ouiaSafe: true,
};

export default FilterDropdown;
