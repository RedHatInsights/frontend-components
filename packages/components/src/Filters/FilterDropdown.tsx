import React, { useState } from 'react';
import { getDefaultOUIAId } from '@patternfly/react-core/';
import { Dropdown, DropdownToggle, Level } from '@patternfly/react-core';
import FilterInput from './FilterInput';
import './filter-dropdown.scss';

export type FilterCategory = {
  title?: string;
  type?: string;
  urlParam?: string;
  values?: {
    label?: string;
    value?: string;
  }[];
};

export interface FilterDropdownProps {
  addFilter?: (filterName?: string, selectedValue?: string | number, type?: string) => void;
  removeFilter?: (filterName?: string, selectedValue?: string | number) => void;
  hideCategories?: [string | undefined];
  filters?: Record<string, string>;
  filterCategories?: FilterCategory[];
  label?: React.ReactNode;
  ouiaId?: string;
  ouiaSafe?: boolean;
}

const ouiaStateId = getDefaultOUIAId('RHI/FilterDropdown');

const FilterDropdown: React.FunctionComponent<FilterDropdownProps> = ({
  hideCategories,
  filters,
  filterCategories,
  label,
  ouiaId = ouiaStateId,
  ouiaSafe = true,
  addFilter,
  removeFilter,
  ...props
}) => {
  const [isOpen, setIsOpen] = useState(false);

  const addRemoveFilters = (selectedValue?: string | number, filterName?: string, type?: string, isChecked?: boolean) => {
    if (type === 'checkbox') {
      isChecked ? addFilter?.(filterName, selectedValue, type) : removeFilter?.(filterName, selectedValue);
    } else {
      addFilter?.(filterName, selectedValue, type);
    }
  };

  return (
    <Dropdown
      className="ins-c-filter__dropdown"
      {...props}
      toggle={
        <DropdownToggle ouiaId={ouiaId} ouiaSafe={ouiaSafe} onToggle={(isOpen) => setIsOpen(isOpen)}>
          {label}
        </DropdownToggle>
      }
      isOpen={isOpen}
      ouiaId={ouiaId}
      ouiaSafe={ouiaSafe}
    >
      <div className="pf-c-dropdown__menu-item">
        {filterCategories?.map(
          (data: FilterCategory, index) =>
            !hideCategories?.includes(data?.urlParam) && (
              <Level key={`${data.urlParam}${index}`}>
                {data.title || ''}
                {data.values?.map((item: { label?: React.ReactNode; value?: string }, key: string | number) => (
                  <FilterInput
                    key={`check${index}${key}`}
                    aria-label={item.label}
                    id={`${data.urlParam}${key}`}
                    label={item.label}
                    addRemoveFilters={addRemoveFilters}
                    param={data.urlParam}
                    type={data.type}
                    value={item.value}
                    filters={filters}
                  />
                ))}
                {index !== filterCategories.length - 1 && <br />}
              </Level>
            )
        )}
      </div>
    </Dropdown>
  );
};

export default FilterDropdown;
