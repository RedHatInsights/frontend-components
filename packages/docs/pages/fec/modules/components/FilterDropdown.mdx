
# FilterDropdown

## Props

|name|type|default|description|
|----|----|-------|-----------|
|addFilter|(filterName?:string,selectedValue?:string&#124;number,type?:string)=&gt;void|||
|removeFilter|(filterName?:string,selectedValue?:string&#124;number)=&gt;void|||
|hideCategories|[string&#124;undefined]|||
|filters|Record&lt;string,string&gt;|||
|filterCategories|FilterCategory[]|||
|label|React.ReactNode|||
|ouiaId|`string`|`getDefaultOUIAId('RHI/FilterDropdown')`||
|ouiaSafe|<code>&#123;<br/><span className="default-prop-spacer"></span><span className="default-prop-spacer"></span><span className="default-prop-spacer"></span> "name": "boolean"<br />&#125;</code>|`true`||



