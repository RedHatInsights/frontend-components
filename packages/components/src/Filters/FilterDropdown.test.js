import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import FilterDropdown from './FilterDropdown';

describe('FilterDropdown component', () => {
    const defaultProps = {
        filters: {},
        addFilter: jest.fn(),
        removeFilter: jest.fn(),
        filterCategories: [
            { title: '', type: '', urlParam: '', values: [{ label: '', value: '' }] }
        ]
    };

    it('should render', () => {
        const wrapper = shallow(
            <FilterDropdown { ...defaultProps }/>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render with a component as label', () => {
        const wrapper = shallow(
            <FilterDropdown { ...defaultProps } label={ <React.Fragment /> }/>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
