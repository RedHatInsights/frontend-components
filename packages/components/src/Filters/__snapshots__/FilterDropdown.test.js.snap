// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterDropdown component should render 1`] = `
<Dropdown
  className="ins-c-filter__dropdown"
  isOpen={false}
  ouiaId="OUIA-Generated-RHI/FilterDropdown-1"
  ouiaSafe={true}
  toggle={
    <DropdownToggle
      onToggle={[Function]}
      ouiaId="OUIA-Generated-RHI/FilterDropdown-1"
      ouiaSafe={true}
    />
  }
>
  <div
    className="pf-c-dropdown__menu-item"
  >
    <Level
      key="0"
    >
      <FilterInput
        addRemoveFilters={[Function]}
        aria-label=""
        filters={Object {}}
        id="0"
        key="check00"
        label=""
        param=""
        type=""
        value=""
      />
    </Level>
  </div>
</Dropdown>
`;

exports[`FilterDropdown component should render with a component as label 1`] = `
<Dropdown
  className="ins-c-filter__dropdown"
  isOpen={false}
  ouiaId="OUIA-Generated-RHI/FilterDropdown-1"
  ouiaSafe={true}
  toggle={
    <DropdownToggle
      onToggle={[Function]}
      ouiaId="OUIA-Generated-RHI/FilterDropdown-1"
      ouiaSafe={true}
    >
      <React.Fragment />
    </DropdownToggle>
  }
>
  <div
    className="pf-c-dropdown__menu-item"
  >
    <Level
      key="0"
    >
      <FilterInput
        addRemoveFilters={[Function]}
        aria-label=""
        filters={Object {}}
        id="0"
        key="check00"
        label=""
        param=""
        type=""
        value=""
      />
    </Level>
  </div>
</Dropdown>
`;
